### PR TITLE
Write msyqlType to canal packets and add special handling for Blob & Text field types.

### PR DESCRIPTION
**Pull request by qinggniq**
### What problem does this PR solve? 
1. #697 add mysqlType to canal column entry.
2. Fix text/blob type issue.

### What is changed and how it works?
1. Add `Charset` field to `mode.Column` to help figure out real mysqlType.
2. Now in cdc `TEXT` and `BLOB` mysql types are both mapped to `JavaSQLTypeBLOB`, but in [jdbc](https://dev.mysql.com/doc/connector-j/5.1/en/connector-j-reference-type-conversions.html), `TEXT` is mapped to `JavaSQLTypeVARCHAR`, `BLOB` is mapped to `JavaSQLTypeBLOB`, so canal consumer (alibaba canal consumer/adapter) will use different logic to handle `TEXT` and `BLOB` type values, but the encoding logic for `TEXT` and `BLOB` in cdc is same, there might be something wrong. So we should distinguish them. 

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported variable/fields change

### CODE REVIEW

File: cdc/entry/mounter.go [hunk 1]
```diff
@@ -373,8 +373,9 @@ func (m *mounterImpl) mountRowKVEntry(tableInfo *TableInfo, row *rowKVEntry) (*m
 			return nil, errors.Trace(err)
 		}
 		col := &model.Column{
-			Type:  colInfo.Tp,
-			Value: value,
+			Type:    colInfo.Tp,
+			Value:   value,
+			Charset: colInfo.Charset,
```
Comment (liuzix) [inline cdc/entry/mounter.go:378]: Please use a flag to indicate whether it is binary data or not.

File: cdc/model/sink.go [hunk 2]
```diff
@@ -87,6 +88,7 @@ type RowChangedEvent struct {
 type Column struct {
 	Type        byte        `json:"t"`
 	WhereHandle *bool       `json:"h,omitempty"`
+	Charset     string      `json:"c"`
```
Comment (liuzix) [inline cdc/model/sink.go:91]: Probably should use int32 or int64 for the sake of extensibility. We may need to store other flags in the future.

File: cdc/sink/codec/canal.go [hunk 3]
```diff
@@ -126,6 +129,7 @@ func (b *canalEntryBuilder) buildHeader(commitTs uint64, schema string, table st
 // build the Column in the canal RowData
 func (b *canalEntryBuilder) buildColumn(c *model.Column, colName string, updated bool) (*canal.Column, error) {
 	sqlType := MysqlToJavaType(c.Type)
+	mysqlType := parser_types.TypeToStr(c.Type, c.Charset)
```
Comment (liuzix) [inline cdc/sink/codec/canal.go:132]: You only need to determine whether Charset is 'binary' or 'utf8'.

File: cdc/sink/codec/canal.go [hunk 4]
```diff
@@ -163,12 +173,18 @@ func (b *canalEntryBuilder) buildColumn(c *model.Column, colName string, updated
 		case string:
 			value = v
 		case []byte:
-			decoded, err := b.bytesDecoder.Bytes(v)
-			if err != nil {
-				return nil, errors.Trace(err)
+			// special handle for text and blob
+			// see https://github.com/alibaba/canal/blob/9f6021cf36f78cc8ac853dcf37a1769f359b868b/parse/src/main/java/com/alibaba/otter/canal/parse/inbound/mysql/dbsync/LogEventConvert.java#L801
+			if strings.Contains(mysqlType, "text") {
```
Comment (liuzix) [inline cdc/sink/codec/canal.go:178]: Don't need to test it again. Just use sqlType.

File: cdc/sink/codec/canal.go (hunk 3 above)
Comment (qinggniq) [reply cdc/sink/codec/canal.go:132]: You means that we don't need mysqlType in canal column now?

File: cdc/model/sink.go (hunk 2 above)
Comment (qinggniq) [reply cdc/model/sink.go:91]: use `0x0` for binary charset,`0x01` for others?

File: cdc/sink/codec/canal.go [hunk 5]
```diff
@@ -148,12 +155,12 @@ func (b *canalEntryBuilder) buildColumn(c *model.Column, colName string, updated
 	}
 	switch sqlType {
 	case JavaSQLTypeBINARY, JavaSQLTypeVARBINARY, JavaSQLTypeLONGVARBINARY:
-		if strings.Contains(mysqlType, "text") {
+		if c.Flag == model.BinaryFlag {
```
[the remark this answers is not in the document]
Comment (qinggniq) [reply cdc/sink/codec/canal.go:158]: ok

File: cdc/entry/mounter.go [hunk 6]
```diff
@@ -358,6 +358,13 @@ func (m *mounterImpl) mountRowKVEntry(tableInfo *TableInfo, row *rowKVEntry) (*m
 	if !row.Delete {
 		datumsNum = len(tableInfo.Columns)
 	}
+
+	charsetFlag := func(charset string) uint32 {
```
[the remark this answers is not in the document]
Comment (qinggniq) [reply cdc/entry/mounter.go:362]: No, just there are duplicate codes.

File: cdc/model/sink.go [hunk 7]
```diff
@@ -85,9 +92,10 @@ type RowChangedEvent struct {
 
 // Column represents a column value in row changed event
 type Column struct {
-	Type        byte        `json:"t"`
-	WhereHandle *bool       `json:"h,omitempty"`
-	Value       interface{} `json:"v"`
+	Type        byte         `json:"t"`
+	WhereHandle *bool        `json:"h,omitempty"`
+	Flag        bitflag.Flag `json:"f"`
```
Comment (leoppro) [inline cdc/model/sink.go:97]: could you add the unit test for this bitflag?

Comment (qinggniq) [reply cdc/model/sink.go:97]: OK, I will implement it at pkg/util/bitflag.go, and use pingcap check to write some uint tests.

Comment (leoppro) [reply cdc/model/sink.go:97]: and I found that the `bitflag.Flag` is alias of `byte`, so the `bitflag.Flag` can store up to 8 bits of flags. It seems that 8 bits are not enough, could we using `uint64` to store the flags?

Comment (qinggniq) [reply cdc/model/sink.go:97]: Sure!

File: cdc/model/sink.go [hunk 8]
```diff
@@ -34,6 +37,29 @@ const (
 	MqMessageTypeResolved
 )
 
+// ColumnFlagType is for encapsulating the flag operations for different flags.
+type ColumnFlagType util.Flag
+
+const (
+	// BinaryFlag means col charset is binary
+	BinaryFlag ColumnFlagType = 1 << 0
```
Comment (liuzix) [inline cdc/model/sink.go:45]: Please use iota

File: cdc/model/sink.go [hunk 9]
```diff
@@ -34,6 +37,29 @@ const (
...
+)
+
+//SetIsBinary set BinaryFlag
+func (b *ColumnFlagType) SetIsBinary() {
+	(*util.Flag)(b).Add(util.Flag(BinaryFlag))
+}
+
+//UnSetIsBinary unset BinaryFlag
+func (b *ColumnFlagType) UnSetIsBinary() {
```
Comment (liuzix) [inline cdc/model/sink.go:54]:
```suggestion
func (b *ColumnFlagType) UnsetIsBinary() {
```

File: cdc/model/sink.go [hunk 10]
```diff
@@ -34,6 +37,29 @@ const (
...
+	(*util.Flag)(b).Remove(util.Flag(BinaryFlag))
+}
+
+//IsSetBinary show whether BinaryFlag is set
+func (b *ColumnFlagType) IsSetBinary() bool {
```
Comment (liuzix) [inline cdc/model/sink.go:59]:
```suggestion
func (b *ColumnFlagType) IsBinary() bool {
```

File: cdc/sink/codec/canal.go [hunk 11]
```diff
@@ -181,13 +205,15 @@ func (b *canalEntryBuilder) buildColumn(c *model.Column, colName string, updated
...
 		Value:         value,
+		MysqlType:     mysqlType,
 	}
 	return canalColumn, nil
 }
 
 // build the RowData of a canal entry
 func (b *canalEntryBuilder) buildRowData(e *model.RowChangedEvent) (*canal.RowData, error) {
 	var columns []*canal.Column
+
```
Comment (liuzix) [inline cdc/sink/codec/canal.go:216]:
```suggestion
```

File: cdc/sink/codec/canal_test.go [hunk 12]
```diff
@@ -147,7 +149,7 @@ func (s *canalEntrySuite) TestConvertEntry(c *check.C) {
 		},
 		Delete: true,
 		Columns: map[string]*model.Column{
-			"id": {Type: mysql.TypeLong, WhereHandle: &trueVar, Value: 1},
+			"id": {Type: mysql.TypeLong, WhereHandle: &trueVar, Value: 1, Flag: model.BinaryFlag},
```
Comment (liuzix) [inline cdc/sink/codec/canal_test.go:152]: Why would you want to set the BinaryFlag on a long type?

Comment (qinggniq) [reply cdc/sink/codec/canal_test.go:152]: I just obey the information from `timodel.ColumnInfo.FieldType.Charset`, all type have their charset, and for `bigint` it shows the `Charset : binary` in my `cdc.log`.

Comment (qinggniq) [reply cdc/sink/codec/canal_test.go:152]: Sorry, it's my illusion.

File: cdc/entry/mounter.go [hunk 13]
```diff
@@ -405,9 +411,14 @@ func (m *mounterImpl) mountRowKVEntry(tableInfo *TableInfo, row *rowKVEntry) (*m
 		for _, col := range tableInfo.Columns {
 			_, ok := values[col.Name.O]
 			if !ok && tableInfo.IsColWritable(col) {
+				var flag model.ColumnFlagType
+				if col.Charset == "binary" {
+					flag.SetIsBinary()
+				}
```
Comment (amyangfei) [inline cdc/entry/mounter.go:417]: recommend encapsulating this logic into a function like
```golang
func transColumnFlag(col timodel.ColumnInfo) model.ColumnFlagType {
    // some logic
}
```
we can reuse it

Comment (qinggniq) [reply cdc/entry/mounter.go]: OK.

**Comment by amyangfei** [comment]
Good job!

**Pull request merged**